README.rst

```diff
@@ -1,102 +1,208 @@
+.. _hygienic macros: TODO: write a short example demonstrating lack of capture.
+.. _easy-to-override hygiene: examples/anaphoric-if.kl
+.. _Custom languages: examples/rpn.kl
+.. _Syntax objects: TODO: link to a short example which explains that in
+   Racket, syntax objects are introduced via ``#'(...)``, whereas in Klister
+   they are introduced via ``'(...)``. Also explain that Klister does not have
+   unannotated s-expressions.
+.. _module system: TODO: write a short example demonstrating what can be
+   expressed with the import and export primitives.
+
+Features we borrow from Haskell:
+
+* Statically-typed; we check that macro implementations have type
+  ``(-> Syntax (Macro Syntax)``, and *then* that the generated code is also
+  well-typed. We thus aim for the expressivity of Template Haskell, not the
+  extra guarantees of Typed Template Haskell.
+* Purely functional; primitives with side-effects either run in the ``IO``
+  monad or in the ``Macro`` monad, as appropriate.
+* Higher-kinded types; for example `monads`_ are defined as a library.
+* `Algebraic datatypes`_. In Racket, ``(list 1 2 3)`` and ``'(1 2 3)``
+  construct the same value, an s-expression which satisfies the ``list?``
+  predicate. In Klister, the former produces a value of the algebraic datatype
+  ``(List Integer)``, while the latter produces a value of type ``Syntax``.
+  This syntax can be "opened", via ``(open-syntax '(1 2 3))``, to obtain the
+  value ``(list-contents (list '1 '2 '3))`` of type
+  ``(Syntax-Contents Syntax)``.
+* `Bidirectional type inference`_; the type of an argument can specialize the
+  type of the function being called.
+* Bidirectional type inference; the type of a function can specialize the type
+  of its arguments.
+
+.. _monads: TODO: link to monad.kl's Monad definition, and add a comment there
+   highlighting the inferred type, especially the higher-kinded type variable.
+.. _Algebraic datatypes: TODO: write a small example defining and matching on
+   an algebraic type. Perhaps Either?
+.. _Bidirectional type inference: TODO: write a small example demonstrating
+   those two features.
+
+Features which make Klister special (but not *unique*; see the `bibliography`_):
+
+* `Type-providing macros`_; a macro used in an argument position can provide
+  type information about the argument, and can thus specialize the type of the
+  function being called.
+* `Type-aware macros`_; a macro used in an argument position can obtain the
+  type of the argument it needs to generate, and thus the type of a function
+  can affect what arguments are generated.
+* `Stuck macros`_; it is thus possible for multiple macros to affect what each
+  other generates. The language primitives are designed so that the order in
+  which the macros are expanded cannot affect their results, and indeed the
+  same is true for the order in which the macro expansion and type-inference
+  steps are interleaved. This makes Klister code more robust to refactoring.
```

Comment:
```suggestion
  steps are interleaved. This means that the order in which the type checker traverses a program and generates constraints is not visible to the authors of macros, providing a predictable programming model.
```

Reply:
Hmm, can we keep both? I think the robustness to refactoring is a real and convincing advantage. With a type-aware macros which reveals whatever the type checker has inferred so far, you could easily end up in the following situation. The user writes a macro, and it just so happens that with the current task ordering, the macro has access to all the information it needs and the code works fine. Then, the user makes what seems like a benign refactoring, like moving the macro definition to a separate file or naming an expression with a `let`. This affects the task ordering, and now the macro sees type variables instead of the information it needed, so the code no longer works. That is, the code was not robust enough to survive the refactoring.

Reply:
I think both is good here. I would have interpreted the earlier text as saying that programs in general are easier to refactor, rather than just metaprograms, but I think that combining these two texts will make that clear.

Reply:
combined both in c8045298c7f7a94d4195cab17bfb14880ac546f3